dbms/src/Storages/KVStore/MultiRaft/RegionSerde.cpp

```diff
@@ -0,0 +1,279 @@
+
+using MaybeRegionPersistExtension = UInt32;
+enum class RegionPersistExtension : MaybeRegionPersistExtension
+{
+    Reserved1 = 1,
+    ReservedForTest = 2,
+    // It should always be equal to the maximum supported type + 1
+    MaxKnownFlag = 3,
+};
+
+constexpr MaybeRegionPersistExtension UNUSED_EXTENSION_NUMBER_FOR_TEST = UINT32_MAX / 2;
+static_assert(!magic_enum::enum_contains<RegionPersistExtension>(UNUSED_EXTENSION_NUMBER_FOR_TEST));
+static_assert(std::is_same_v<MaybeRegionPersistExtension, UInt32>);
+static_assert(magic_enum::enum_underlying(RegionPersistExtension::MaxKnownFlag) <= UINT32_MAX / 2);
+static_assert(
+    magic_enum::enum_count(RegionPersistExtension)
+    == magic_enum::enum_underlying(RegionPersistExtension::MaxKnownFlag));
+
+constexpr UInt32 Region::CURRENT_VERSION = static_cast<UInt64>(RegionPersistVersion::V2);
+
+std::pair<MaybeRegionPersistExtension, UInt32> getPersistExtensionTypeAndLength(ReadBuffer & buf)
+{
+    auto ext_type = readBinary2<MaybeRegionPersistExtension>(buf);
+    auto size = readBinary2<UInt32>(buf);
+    // Note `ext_type` may not valid in RegionPersistExtension
+    return std::make_pair(ext_type, size);
+}
+
+size_t writePersistExtension(
+    UInt32 & cnt,
+    WriteBuffer & wb,
+    MaybeRegionPersistExtension ext_type,
+    const char * data,
+    UInt32 size)
+{
+    auto total_size = writeBinary2(ext_type, wb);
+    total_size += writeBinary2(size, wb);
+    wb.write(data, size);
+    total_size += size;
+    cnt++;
+    return total_size;
+}
+
+std::tuple<size_t, UInt64> Region::serialize(WriteBuffer & buf) const
+{
+    auto binary_version = Region::CURRENT_VERSION;
+    // Increase this when persist with a new extension type.
+    UInt32 expected_extension_count = 0;
+    using bundle_type = std::pair<int, int>;
+    fiu_do_on(FailPoints::force_region_persist_version, {
+        if (auto v = FailPointHelper::getFailPointVal(FailPoints::force_region_persist_version); v)
+        {
+            std::tie(binary_version, expected_extension_count) = std::any_cast<bundle_type>(v.value());
+            LOG_WARNING(
+                Logger::get(),
+                "Failpoint force_region_persist_version set region binary version, value={}, "
+                "expected_extension_count={}",
+                binary_version,
+                expected_extension_count);
+        }
+    });
+    size_t total_size = writeBinary2(binary_version, buf);
+    UInt64 applied_index = -1;
+
+    {
+        std::shared_lock<std::shared_mutex> lock(mutex);
+
+        // Serialize meta
+        const auto [meta_size, index] = meta.serialize(buf);
+        total_size += meta_size;
+        applied_index = index;
+
+        // Try serialize extra flags
+        if (binary_version >= 2)
+        {
+            static_assert(sizeof(eager_truncated_index) == sizeof(UInt64));
+            UInt32 flags = RegionPersistFormat::HAS_EAGER_TRUNCATE_INDEX;
+            flags |= (expected_extension_count << 1);
```

Comment:
I perfer

```suggestion
            // The upper 31 bits are used to store the length of extensions, and the lowest bit is flag of eager gc.
            UInt32 flags = (expected_extension_count << 1) + 1;
```

Reply:
I think `(expected_extension_count << 1) | RegionPersistFormat::HAS_EAGER_TRUNCATE_INDEX` is better.

Reply:
The better add some more obvious comments, now the code makes me confused.

Reply:
How about I assert `static_assert(RegionPersistFormat::HAS_EAGER_TRUNCATE_INDEX == 0x01);` here?